addons/resource/models/resource.py

```diff
@@ -987,6 +986,129 @@ def _get_unavailable_intervals(self, start, end):
+
+            work_intervals = [(start, stop) for start, stop, meta in resource_work_intervals]
+            # start + flatten(intervals) + end
+            work_intervals = [start_datetime] + list(chain.from_iterable(work_intervals)) + [end_datetime]
+            # put it back to UTC
+            work_intervals = list(map(lambda dt: dt.astimezone(utc), work_intervals))
+            # pick groups of two
+            work_intervals = list(zip(work_intervals[0::2], work_intervals[1::2]))
+            leaves_mapping[res.id] = work_intervals
+        return leaves_mapping
+
+    def _personal_leaves_intervals(self, start_dt, end_dt):
+        """
+        :return: a dict { res.id : list[date_start, date_stop] } with the personal leaves of each resource
+        """
+        domain = [
+            ('time_type', '=', 'leave'),
+            ('resource_id', 'in', self.ids),
+            ('date_from', '<=', end_dt),
+            ('date_to', '>=', start_dt),
+        ]
+        leaves = self.env['resource.calendar.leaves'].search_read(domain, fields=['date_from', 'date_to', 'resource_id'])
+        dict_personal_leaves = {}
+        for leave in leaves:
+            if dict_personal_leaves.get(leave['resource_id'][0]):
+                dict_personal_leaves[leave['resource_id'][0]].append([leave['date_from'], leave['date_to']])
+            else:
+                dict_personal_leaves[leave['resource_id'][0]] = [[leave['date_from'], leave['date_to']]]
+        return dict_personal_leaves
+
+    def _global_leaves_per_calendar(self, start_dt, end_dt, calendars_set):
+        """
+        :return: a dict { res.id : list[date_start, date_stop] } with the global leaves of each calendar. Each dict entry also contains the global leaves that have
+        no calendar_id
+        """
+        domain = [
+            ('resource_id', '=', False),
+            ('time_type', '=', 'leave'),
+            ('calendar_id', 'in', [False] + calendars_set.ids),
+            ('date_from', '<=', end_dt),
+            ('date_to', '>=', start_dt),
+        ]
+        leaves = self.env['resource.calendar.leaves'].search_read(domain, fields=['date_from', 'date_to', 'calendar_id'])
+        dict_global_leaves = {}
+        for leave in leaves:
+            key = leave['calendar_id'][0] if leave['calendar_id'] else False # the 'if' is needed in case of no calendar_id. It avoids a traceback on 'leave['calendar_id'][0]'
+            if dict_global_leaves.get(key):
+                dict_global_leaves[key].append([leave['date_from'], leave['date_to']])
+            else:
+                dict_global_leaves[key] = [[leave['date_from'], leave['date_to']]]
+        # in case some global leaves are not linked to any calendar, we must add them to each dict entry to ensures they are computed for each user
+        if not dict_global_leaves.get(False):
+            return dict_global_leaves
+        common_leaves = dict_global_leaves[False]
+        for key in dict_global_leaves:
```

Comment:
And this could be done in the previous loop, right? :slightly_smiling_face:

Reply:
I dont think it's doable. Say for example, you have 2 leave in the recordset, and you iterate through it. The first is a leave with no calendar_id, the second is  a leave with calendar_id = 2. 
The iteration will first create an entry for the key 'False', but it wont be possible to add it to the key  '2' since it has not been created yet.

Reply:
Nvm, you are right. :see_no_evil: